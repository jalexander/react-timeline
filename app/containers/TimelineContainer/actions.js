/*
 *
 * TimelineContainer actions
 *
 */

import {
  REQUEST_TIMELINE,
  REQUEST_TIMELINE_SUCCEEDED,
  REQUEST_TIMELINE_FAILED,
} from './constants';

export function requestTimeline() {
  return {
    type: REQUEST_TIMELINE,
  };
}

export function requestTimelineSucceeded(timeline) {
  return {
    type: REQUEST_TIMELINE_SUCCEEDED,
    timeline,
  };
}

export function requestTimelineFailed(message) {
  return {
    type: REQUEST_TIMELINE_FAILED,
    message,
  };
}
