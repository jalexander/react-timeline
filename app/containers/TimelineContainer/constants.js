/*
 *
 * TimelineContainer constants
 *
 */

export const REQUEST_TIMELINE = 'app/TimelineContainer/REQUEST_TIMELINE';
export const REQUEST_TIMELINE_SUCCEEDED = 'app/TimelineContainer/REQUEST_TIMELINE_SUCCEEDED';
export const REQUEST_TIMELINE_FAILED = 'app/TimelineContainer/REQUEST_TIMELINE_FAILED';
export const SET_ACTIVE_MARKER = 'app/TimelineContainer/SET_ACTIVE_MARKER'
