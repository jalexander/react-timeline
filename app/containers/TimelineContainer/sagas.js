import { call, put, takeLatest } from 'redux-saga/effects';
import { REQUEST_TIMELINE } from './constants';
import { requestTimelineSucceeded, requestTimelineFailed } from './actions';

export function fetchTimelineFromServer() {
  return fetch('http://localhost:3000/api/timeline')
    .then((response) => response.json());
}

export function* fetchTimeline() {
  try {
    const timeline = yield call(fetchTimelineFromServer);
    yield put(requestTimelineSucceeded(timeline));
  } catch (error) {
    yield put(requestTimelineFailed(error));
  }
}

// Individual exports for testing
export function* fetchTimelineSaga() {
  yield takeLatest(REQUEST_TIMELINE, fetchTimeline);
}

// All sagas to be loaded
export default [
  fetchTimelineSaga,
];
