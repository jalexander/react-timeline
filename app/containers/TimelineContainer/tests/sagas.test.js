/**
 * Test  sagas
 */

import { put, takeLatest } from 'redux-saga/effects';

import { REQUEST_TIMELINE } from '../constants';
import { requestTimelineSucceeded, requestTimelineFailed } from '../actions';

import { fetchTimelineSaga, fetchTimeline } from '../sagas';

/* eslint-disable redux-saga/yield-effects */
describe('fetchTimeline Saga', () => {
  let fetchTimelineGenerator;

  beforeEach(() => {
    fetchTimelineGenerator = fetchTimeline();

    const callDescriptor = fetchTimelineGenerator.next().value;
    expect(callDescriptor).toMatchSnapshot();
  });

  it('should dispatch the requestTimelineSucceeded action if it requests the data successfully', () => {
    const response = [{
      name: 'First repo',
    }, {
      name: 'Second repo',
    }];
    const putDescriptor = fetchTimelineGenerator.next(response).value;
    expect(putDescriptor).toEqual(put(requestTimelineSucceeded(response)));
  });

  it('should call the repoLoadingError action if the response errors', () => {
    const response = new Error('Some error');
    const putDescriptor = fetchTimelineGenerator.throw(response).value;
    expect(putDescriptor).toEqual(put(requestTimelineFailed(response)));
  });
});

describe('fetchTimelineSaga Saga', () => {
  const fetchTimelineSagaGenerator = fetchTimelineSaga();

  it('should start task to watch for REQUEST_TIMELINE action', () => {
    const takeLatestDescriptor = fetchTimelineSagaGenerator.next().value;
    expect(takeLatestDescriptor).toEqual(takeLatest(REQUEST_TIMELINE, fetchTimeline));
  });
});
