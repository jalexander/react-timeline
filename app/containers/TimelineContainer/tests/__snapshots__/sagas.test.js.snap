exports[`fetchTimeline Saga should call the repoLoadingError action if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`fetchTimeline Saga should dispatch the requestTimelineSucceeded action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [],
    "context": null,
    "fn": [Function],
  },
}
`;
